File: 2020.08.21/2020.08.21 Classic.py
#Visual of Hamster Pen.
#
#Rohan Lewis
#2020.08.24

import math
import simplegui

#Globals
HEIGHT = 300
WIDTH = 550
CENTER = (150, 150)
n_sides = 3
post_weight = 0.3333

def reset() :
    global n_sides, post_weight
    n_sides = 3
    post_weight = 0.3333

def draw(canvas):
    global HEIGHT, WIDTH, CENTER, n_sides, post_weight
    
    increment_n()
    
    side = p_side(n_sides, post_weight)
    angle = p_angle(n_sides)
    apothem = p_apothem(side, angle)
    radius = p_radius(side, angle)
    
    area = p_area(apothem, n_sides, side)
    vertices = p_vertices(angle, n_sides, radius)
    
    #Color background white.
    canvas.draw_polygon([(0, 0), (0, HEIGHT), (WIDTH, HEIGHT), (WIDTH, 0)], 2, "Red", "White")
    
    #Draw top line.
    canvas.draw_polygon(vertices, 2, "Blue", "Blue")
    
    #Number of Sides.
    canvas.draw_text('Number of Sides : ' + str(n_sides), [270, 30], 20, 'Black')
    
    #Post Weight.
    if post_weight == 0 :
        canvas.draw_text('Post Weight : ~0 kilograms', [270, 50], 20, 'Black')
    else :
        canvas.draw_text('Post Weight : ' + str(post_weight) + ' kilograms', [270, 50], 20, 'Black') 
        
    #Side Length.
    canvas.draw_text('Side Length : ' + str(round(side, 5)) + ' meters', [270, 70], 20, 'Black')
    
    #Area of Polygon.
    canvas.draw_text('Pen Area : ' + str(round(area, 5)) + ' square meters', [270, 90], 20, 'Black')

    if post_weight > 0 :
        if n_sides == 3 :
            post_weight -= 0.001
        elif n_sides < 7 :
            post_weight -= 0.0001
        elif n_sides < 12 :
            post_weight -= 0.00001
        else : 
            post_weight -= 0.000001
    else :
        post_weight = 0
'''Calculates the minimum weight of a post for side n before
incrementing n.'''
def get_min_weight(n) :
    t_1 = math.tan(math.pi / n)
    t_2 = math.tan(math.pi / (n+1))
    a = n * (n + 1) * (n * t_2 - (n+1) * t_1)
    b = -2 * n * (n + 1) * (t_2 - t_1)
    c = ((n+1) * t_2 - n * t_1)
    d = b ** 2 - 4 * a * c
    min_weight = (-1 * b + math.sqrt(d)) / (2 * a)
    return(min_weight)

'''Increases n by 1 if weight drops below minimum weight.'''
def increment_n() :
    global n_sides
    if post_weight <= get_min_weight(n_sides) :
        n_sides += 1        
        
'''Takes the number of sides and post weight, returns polygon
side length so that total weight is 1 kilogram)'''
def p_side(n, weight) :
    side = (1.0 - n * weight) / n
    return(side)

'''Takes the number of sides and returns the angle between
the apothem and radius.'''            
def p_angle(n) :
    angle = math.pi / n
    return(angle)

'''Takes the side length and angle and returns the apothem
length.''' 
def p_apothem(side, angle) :
    apothem = side / (2 * math.tan(angle))
    return(apothem)

'''Takes the side length and angle and returns the radius
length.''' 
def p_radius(side, angle) :
    radius = side / (2 * math.sin(angle))
    return(radius)

'''Takes the apothem, number of sides, and side length and
returns the area.'''           
def p_area(apothem, n, side) :
    area = apothem * n * side / 2
    return(area)

'''Takes the angle, number of sides, and radius, and returns
the coordinates for the polygon.''' 
def p_vertices(angle, n, radius) :
    global CENTER
    vertices = []
    for v in range(n) :
        x = CENTER[0] + 750 * radius * math.cos(2 * angle * v)
        y = CENTER[1] + 750 * radius * math.sin(2 * angle * v) 
        vertices.append((x,y))

    return(vertices)    

# Create frame.
frame = simplegui.create_frame("Hamster Pen", WIDTH, HEIGHT)
frame.set_draw_handler(draw)
#frame.set_keydown_handler(keydown)
#frame.set_keyup_handler(keyup)

# Start frame.
reset()
frame.start()

File: 2020.08.21/2020.08.21 Classic.R
sides <- NULL
lengths <- NULL
min_weights <- NULL
post_weights <- NULL
areas <- NULL


#Calculates the minimum weight of a post for side n before incrementing n.
get_min_weight <- function(n) {
  t_1 <- tan(pi / n)
  t_2 <- tan(pi / (n+1))
  a <- n * (n + 1) * (n * t_2 - (n+1) * t_1)
  b <- -2 * n * (n + 1) * (t_2 - t_1)
  c <- ((n+1) * t_2 - n * t_1)
  d <- b ** 2 - 4 * a * c
  (-1 * b + sqrt(d)) / (2 * a)
}

#Increases n by 1 if weight drops below minimum weight.
increment_n <- function(n, post_weight, min_weight) {
  if (post_weight <= min_weight) {
    n + 1
  } else {
    n
  }
}

n <- 3
for (post_weight in seq(0.3333, 0, -.0001)) {
  
  min_weight <- get_min_weight(n)
  n <- increment_n(n, post_weight, min_weight)
  
  #Takes the number of sides and post weight, returns polygon side length so
  #that total weight is 1 kilogram).
  length <- (1 - n * post_weight) / n
  
  #Takes the number of sides and returns the angle between the apothem and radius.
  angle <- pi / n
  
  #Takes the side length and angle and return the apothem length.
  apothem <- length / (2 * tan(angle))
  
  #Takes the side length and angle and returns the radius length. 
  radius <- length / (2 * sin(angle))

  #Takes the apothem, number of sides, and side length and returns the area.           
  area <- apothem * n * length / 2
  
  sides <- c(sides, n)
  lengths <- c(lengths, length)
  min_weights <- c(min_weights, min_weight)
  post_weights <- c(post_weights, post_weight)
  areas <- c(areas, area)
}

min_weights <- c(min_weights, get_min_weight(n))

polygons <- data.frame("Sides" = sides,
                       "Sides2" = ifelse(sides < 9, sides, "More than 8"),
                       "Lengths" = lengths,
                       "Weights" = post_weights,
                       "Areas" = areas)

answer <- data.frame("Sides" = unique(sides),
                     "Min_Weights" = unique(min_weights))


g <- ggplot(polygons, aes(x = Lengths, y = Areas, color = as.factor(Sides2)))
g <- g + geom_point(size = 3)
#Title
g <- g + ggtitle("Maximum Hamster Pen Area vs Length and Number of Sides.")
#X-axis
g <- g + scale_x_continuous("Length of Side")
#Y-axis.
g <- g + scale_y_continuous(name = "Area of Hamster Pen")
#Viridis Color Scheme.
g <- g + scale_color_viridis("Number of Sides", discrete = TRUE)
#Modify labels and text.
g <- g + theme(plot.title = element_text(hjust = 0.5, size = 14, face = "bold"),
               axis.text.x = element_text(size = 12),
               axis.title.x = element_text(size = 14, face = "bold"),
               axis.text.y = element_text(size = 12),
               axis.title.y = element_text(size = 14, face = "bold"),
               legend.text = element_text(size = 12),
               legend.title = element_text(size = 14, face = "bold"))